### ISSUE TITLE
TODO 39: Class-scoped fixture for build_criterion in GIoU invariant tests

### ISSUE THREAD

**Issue by hongthepotato** [hongthepotato/Grounded-Segment-Anything#84]
## Summary

`tests/integration/test_ml_pipeline_cpu.py::TestBuildCriterionGIoUInvariants` rebuilds `build_criterion(num_classes=1)` ~33 times across its 8 tests. `build_criterion` reads YAML and constructs `GroundingDINOCriterion + HungarianMatcher` — non-trivial cost. Hoisting it to a class-scoped fixture would cut ~3s from the CPU integration suite at no behavioral cost.

**Priority:** P3 (cosmetic perf — does not affect correctness).

## Where

[`tests/integration/test_ml_pipeline_cpu.py::TestBuildCriterionGIoUInvariants._giou_loss`](https://github.com/hongthepotato/Grounded-Segment-Anything/blob/agentic/tests/integration/test_ml_pipeline_cpu.py#L561)

```python
class TestBuildCriterionGIoUInvariants:
    @staticmethod
    def _giou_loss(pred_box, gt_box):
        criterion = build_criterion(num_classes=1)   # rebuilt every call
        ...
```

The 8 tests in the class invoke `_giou_loss` from inside parametrized loops + helper methods, totaling ~33 rebuilds.

## Why

- `build_criterion` reads YAML config + constructs `GroundingDINOCriterion + HungarianMatcher`.
- Adds ~3s to `make test` on the CPU integration lane.
- Not flake-inducing, just wasteful — pure setup duplication.

## Fix

Hoist criterion to a class-scoped pytest fixture and pass it into `_giou_loss`:

```python
class TestBuildCriterionGIoUInvariants:
    @pytest.fixture(scope="class")
    def criterion(self):
        return build_criterion(num_classes=1)

    @staticmethod
    def _giou_loss(criterion, pred_box, gt_box):
        outputs = {"pred_logits": torch.full((1, 1, 256), -100.0), "pred_boxes": pred_box.unsqueeze(0)}
        targets = [{"labels": torch.zeros(1, dtype=torch.long), "boxes": gt_box, "token_labels": torch.zeros(1, 256)}]
        return criterion(outputs, targets)["loss_giou"].item()
```

Each test method then takes `criterion` as a parameter:

```python
@pytest.mark.parametrize("side", [0.2, 0.05, 0.01])
def test_self_match_loss_is_zero_across_box_sizes(self, criterion, side):
    box = torch.tensor([[0.5, 0.5, side, side]])
    loss = self._giou_loss(criterion, box, box)
    ...
```

## Acceptance criteria

- [ ] All 8 tests in `TestBuildCriterionGIoUInvariants` use the shared `criterion` fixture
- [ ] `build_criterion` invocation count drops from ~33 to 1 for the class
- [ ] `make test` runtime on CPU integration lane decreases by ~3s
- [ ] No assertion changes — same test logic, same pinned contracts
- [ ] Verified against the rest of the suite (still 2348 passing, no regressions)

## Surfaced via

`/review` and `/ship` adversarial passes on branch [`integration-tests/ml-pipeline-cpu`](https://github.com/hongthepotato/Grounded-Segment-Anything/tree/integration-tests/ml-pipeline-cpu) (2026-05-08). Full TODO entry preserved in [`TODOS.md`](https://github.com/hongthepotato/Grounded-Segment-Anything/blob/integration-tests/ml-pipeline-cpu/TODOS.md) under `### 39`.

## Related

- TODO 36 (P2) — seed `_gdino_outputs`/`_gdino_targets` helpers
- TODO 38 (P3) — fp16-safe clamp threshold in `ml_engine/utils/box_ops.py`
- TODO 40 (P3) — pin remaining 6 dual-accept `pytest.raises` sites

**Comment by hongthepotato** [comment]
related issues:
#85 #86 #87
